## try语句常用知识

try/except结构。它的基础语法如下：

```python
def safe_int(value):
    """尝试把输入转换为整数"""
    try:
        return int(value)
    except TypeError:
        # 当某类异常被抛出时，将会执行对应 except 下的语句
        print(f'type error: {type(value)} is invalid')
    except ValueError:
        # 你可以在一个 try 语句块下写多个 except
        print(f'value error: {value} is invalid')
    finally:
        # finally 里的语句，无论如何都会被执行，哪怕已经执行了return
        print('function completed')
```

在编写try/except语句时，有几个常用的知识点。

01．把更精确的except语句放在前面



02．使用else分支

```python
try:
    sync_profile(user.profile, to_external=True)
except Exception as e:
    print("Error while syncing user profile")
else:
    send_notification(user, 'profile sync succeeded')
```



03．使用空raise语句

在处理异常时，有时我们可能仅仅想记录下某个异常，然后把它重新抛出，交由上层处理。这时，不带任何参数的raise语句可以派上用场：

```python
def incr_by_key(d, key):
    try:
        d[key] += 1
    except KeyError:
        print(f'key {key} does not exists, re-raise the exception')
        raise
```

##  使用上下文管理器

```python
#使用with 打开文件，文件描述符会在作用域结束后自动被释放
with open('foo.txt') as fp:
    content = fp.read()
```

ith是一个神奇的关键字，它可以在代码中开辟一段由它管理的上下文，并控制程序在进入和退出这段上下文时的行为。比如在上面的代码里，这段上下文所附加的主要行为就是：进入时打开某个文件并返回文件对象，退出时关闭该文件对象。



并非所有对象都能像open('foo.txt')一样配合with使用，只有满足上下文管理器（context manager）协议的对象才行。



上下文管理器是一种定义了“进入”和“退出”动作的特殊对象。要创建一个上下文管理器，只要实现__enter__和__exit__两个魔法方法即可。

```py
class DummyContext:
    def __init__(self, name):
        self.name = name

    def __enter__(self):
        # __enter__会在进入管理器时被调用，同时可以返回结果
        # 这个结果可以通过 as 关键字被调用方获取
        #
        # 此处返回一个增加了随机后缀的name
        return f'{self.name}-{random.random()}'

    def __exit__(self, exc_type, exc_val, exc_tb):
        # __exit__会在退出管理器时被调用
        print('Exiting DummyContext')
        return False
```

它的执行效果如下：

```python
>>> with DummyContext('foo') as name:
...     print(f'Name: {name}')
...
Name: foo-0.021691996029607252
Exiting DummyContext
```

上下文管理器功能强大、用处很多，其中最常见的用处之一，就是简化异常处理工作。

01．用于替代finally语句清理资源

在编写try语句时，finally关键字经常用来做一些资源清理类工作，比如关闭已创建的网络连接：

```python
conn = create_conn(host, port, timeout=None)
try:
    conn.send_text('Hello, world!')
except Exception as e:
    print(f'Unable to use connection: {e}')
finally:
    conn.close()
```

可以用上下文管理器来替代finally语句。做起来很简单，只要在__exit__里增加需要的回收语句即可：

```python
class create_conn_obj:
    """创建连接对象，并在退出上下文时自动关闭"""
    def __init__(self, host, port, timeout=None):
        self.conn = create_conn(host, port, timeout=timeout)
    def __enter__(self):
        return self.conn
    def __exit__(self, exc_type, exc_value, traceback):
        # __exit__会在管理器退出时调用
        self.conn.close()
        return False
```

使用create_conn_obj可以创建会自动关闭的连接对象：

```python
#使用上下文管理器创建连接
with create_conn_obj(host, port, timeout=None) as conn:
    try:
        conn.send_text('Hello, world!')
    except Exception as e:
        print(f'Unable to use connection: {e}')
```



02．用于忽略异常

在执行某些操作时，有时程序会抛出一些不影响正常执行逻辑的异常。

打个比方，当你在关闭某个连接时，假如它已经是关闭状态了，解释器就会抛出AlreadyClosedError异常。这时，为了让程序正常运行下去，你必须用try语句来捕获并忽略这个异常：

```python
try:
    close_conn(conn)
except AlreadyClosedError:
    pass
```

虽然这样的代码很简单，但没法复用。当项目中有很多地方要忽略这类异常时，这些try/except语句就会分布在各个角落，看上去非常凌乱。

如果使用上下文管理器，我们可以很方便地实现可复用的“忽略异常”功能——只要在__exit__方法里稍微写几行代码就行：

```python
class ignore_closed:
    """忽略已经关闭的连接"""
    def __enter__(self):
        pass
    def __exit__(self, exc_type, exc_value, traceback):
        if exc_type == AlreadyClosedError:
            return True
        return False
```

当你想忽略AlreadyClosedError异常时，只要把代码用with语句包裹起来即可：

```python
with ignore_closed():
    close_conn(conn)
```



03．使用contextmanager装饰器

虽然上下文管理器很好用，但定义一个符合协议的管理器对象其实挺麻烦的——得首先创建一个类，然后实现好几个魔法方法。为了简化这部分工作，Python提供了一个非常好用的工具：@contextmanager装饰器。

@contextmanager位于内置模块contextlib下，它可以把任何一个生成器函数直接转换为一个上下文管理器。

```python
from contextlib import contextmanager
@contextmanager
def create_conn_obj(host, port, timeout=None):
    """创建连接对象，并在退出上下文时自动关闭"""
    conn = create_conn(host, port, timeout=timeout)
    try:
        yield conn 
    finally: 
        conn.close()
```

- 以yield关键字为界，yield前的逻辑会在进入管理器时执行（类似于__enter__），yield后的逻辑会在退出管理器时执行（类似于__exit__）
- 如果要在上下文管理器内处理异常，必须用try语句块包裹yield语句



永远只捕获那些可能会抛出异常的语句块；

尽量只捕获精确的异常类型，而不是模糊的Exception；

如果出现了预期外的异常，让程序早点儿崩溃也未必是件坏事。

让模块只抛出与当前抽象级别一致的异常；在必要的地方进行异常包装与转换。



以下是本章要点知识总结。

（1）基础知识

一个try语句支持多个except子句，但请记得把更精确的异常类放在前面

try语句的else分支会在没有异常时执行，因此它可用来替代标记变量

不带任何参数的raise语句会重复抛出当前异常

上下文管理器经常用来处理异常，它最常见的用途是替代finally子句

上下文管理器可以用来忽略某段代码里的异常

使用@contextmanager装饰器可以轻松定义上下文管理器



（2）错误处理与参数校验

当你可以选择编写条件判断或异常捕获时，优先选异常捕获（EAFP）

不要让函数返回错误信息，直接抛出自定义异常吧

手动校验数据合法性非常烦琐，尽量使用专业模块来做这件事

不要使用assert来做参数校验，用raise替代它

处理错误需要付出额外成本，假如能通过设计避免它就再好不过了

在设计API时，需要慎重考虑是否真的有必要抛出错误

使用“空对象模式”能免去一些针对边界情况的错误处理工作



（3）当你捕获异常时：

过于模糊和宽泛的异常捕获可能会让程序免于崩溃，但也可能会带来更大的麻烦

异常捕获贵在精确，只捕获可能抛出异常的语句，只捕获可能的异常类型

有时候，让程序提早崩溃未必是什么坏事

完全忽略异常是风险非常高的行为，大多数情况下，至少记录一条错误日志



（4）当你抛出异常时：

保证模块内抛出的异常与模块自身的抽象级别一致

如果异常的抽象级别过高，把它替换为更低级的新异常

如果异常的抽象级别过低，把它包装成更高级的异常，然后重新抛出

不要让调用方用字符串匹配来判断异常种类，尽量提供可区分的异常